topics/data-science/tutorials/python-flow/tutorial.md

```diff
@@ -0,0 +1,476 @@
+> 4.      # ..
+> 5.  elif z:
+> 6.      # ..
+> 7.  if q:
+> 8.      # ..
+> 9.  else:
+> 10.     # ..
+> 11. elif t:
+> 12.     # ..
+> 13. else e:
+> 14.     # ..
+> ```
+>
+> > ### {% icon solution %} Solution
+> > "Blocks" of if/elif/elses
+> > - must start with an `if`
+> > - can have 0 or more `elif` conditions
+> > - can have 0 or 1 `else` condition (if no else condition is supplied, it's equivalent to `else: <nothing>`)
+> >
+> > The above blocks are parsed together, you could not insert a `print` anywhere within the blocks, but between the blocks it would work.
+> >
+> > - 1-2, Just an `if` by itself. There's no elif, or else, so that's the end of that block
+> > - 3-6, `if` and `elif` get evaluated, there is no `else`, so that's the end of that block
+> > - 7-10, `if` and `else` is fine
+> > - 11-14, error! This is missing an `if` case, it will fail with a syntaxerror.
+> {: .solution}
+{: .question}
+
+```python
+# Test code here.
+```
+
+> ### {% icon question %} Question: How Many Paths?
+>
+> Consider this code:
+>
+> ```
+> if 4 > 5:
+>     print('A')
+> elif 4 == 5:
+>     print('B')
+> elif 4 < 5:
+>     print('C')
+> ```
+>
+> Which of the following would be printed if you were to run this code?
+> Why did you pick this answer?
+>
+> 1.  A
+> 2.  B
+> 3.  C
+> 4.  B and C
+>
+> > ### {% icon solution %} Solution
+> > C gets printed because the first two conditions, `4 > 5` and `4 == 5`, are not true,
+> > but `4 < 5` is true.
+> {: .solution}
+{: .question}
+
+```python
+# Test code here.
+```
+
+> ### {% icon question %} Question: Close Enough
+>
+> Write some conditions that print `True` if the variable `a` is within `10` of the variable `b`
+> and `False` otherwise.
+> Compare your implementation with your partner's:
+> do you get the same answer for all possible pairs of numbers?
+>
+> > ### {% icon tip %} Tip: abs
+> > There is a [built-in function `abs`][abs-function] that returns the absolute value of
+> > a number:
+> > ```python
+> > print(abs(-12))
+> > ```
+> >
+> > ```
+> > 12
+> > ```
+> {: .tip}
+>
+> > ### {% icon solution %} Solution 1
+> > ```
+> > a = 5
+> > b = 5.1
+> >
+> > if abs(a - b) <= 10:
+> >     print('True')
+> > else:
+> >     print('False')
+> > ```
+> {: .solution}
+>
+> > ### {% icon solution %} Solution 2
+> > ```
+> > print(abs(a - b) <= 10)
+> > ```
+> >
+> > This works because the Booleans `True` and `False`
+> > have string representations which can be printed.
+> {: .solution}
+{: .question}
+
+```python
+# Test code here.
+```
+
+> ### {% icon question %} Question: Pitfalls
+>
+> A *integer* number between 0 and 100 will be provided to this function. Answer these two questions:
+>
+> - Will it always print something? If not, which value(s) fail?
+> - Can you find any numbers the programmer explicitly wanted to handle, that aren't handled as expected?
+>
+> ```
+> num = 42 # Randomly chosen so the code will execute, try changing it around.
+>
+> if num > 90:
+>     print("great score")
```

Comment:
for some reason this shows up untapped on cocalc

Reply:
for us to remember later: it's due to the blank line.

CoCalc uses an unbelievable markdown parser where even when the above block is in a <pre>, it decides that the newline means it should insert a <p> and start a new paragraph. I'm at a loss for words. 

I don't have a good way to detect that, we'll just have to do it by hand :(